File: python/973-K Closest Points to Origin.py
from typing import List
import heapq
import random

class Solution:
    def kClosest(self, points: List[List[int]], K: int) -> List[List[int]]:
        "MinHeap, Time: O(n+klogn), Space: O(n)"
        min_heap = [(x*x+y*y, (x, y)) for x, y in points] # (dist, point)
        heapq.heapify(min_heap) # O(n)
        
        ans = []
        while(K):
            K -= 1
            ans.append(heapq.heappop(min_heap)[1])
        return ans
    def kClosest(self, points: List[List[int]], K: int) -> List[List[int]]:
        "MaxHeap, Time: O(nlogk), Space: O(k)"
        max_heap = [] # (-dist, point)
        for point in points:
            x, y = point
            heapq.heappush(max_heap, (-(x**2+y**2), point))
            if len(max_heap) > K:
                heapq.heappop(max_heap)
        return [point for _, point in max_heap]
    def kClosest(self, points: List[List[int]], K: int) -> List[List[int]]:
        "Quick Select, Time: O(n), Space: O(1)"
        dist = lambda point: point[0]*point[0] + point[1]*point[1]
        def partition(data, left, right):
            i = left
            j = right
            while i != j:                  
                while dist(data[j]) > dist(data[left]) and i < j:
                    j -= 1
                while dist(data[i]) <= dist(data[left]) and i < j:
                    i += 1
                if i < j:
                    data[i], data[j] = data[j], data[i] 
            data[left], data[i] = data[i], data[left]
            return i
        def helper(data, left, right, k):
            if left >= right: return

            pivot = random.randint(left, right)
            data[left], data[pivot] = data[pivot], data[left]

            pivot_idx = partition(data, left, right)
            if(pivot_idx == k): return
            elif(pivot_idx > k): helper(data, left, pivot_idx-1, k)
            else: helper(data, pivot_idx+1, right, k)

        helper(points, 0, len(points)-1, K-1)
        return points[:K]

points = [[1,3],[-2,2]]
K = 1
result = Solution().kClosest(points, K)
print(result) # [[-2,2]]

points = [[3,3],[5,-1],[-2,4]]
K = 2
result = Solution().kClosest(points, K)
print(result) # [[3,3],[-2,4]]

File: cpp/973. K Closest Points to Origin.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

// class Solution {
// // MaxHeap, Time: O(nlogk), Space: O(k)
// public:
//     vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
//         auto comp = [](vi &lhs, vi &rhs) {
//             return lhs[0]*lhs[0]+lhs[1]*lhs[1] <= rhs[0]*rhs[0]+rhs[1]*rhs[1];
//         };
//         priority_queue<vi, vvi, decltype(comp)> pq(comp); 

//         for(vi point : points) {
//             pq.push(point);
//             if(pq.size() > K) pq.pop();
//         }

//         vvi ans;
//         while(pq.size()) {
//             ans.push_back(pq.top()); 
//             pq.pop();
//         }
//         return ans;
//     }
// };

// class Solution {
// // MinHeap, Time: O(n+klogn), Space: O(n)
// public:
//     vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
//         auto comp = [](vi &lhs, vi &rhs) {
//             return lhs[0]*lhs[0]+lhs[1]*lhs[1] > rhs[0]*rhs[0]+rhs[1]*rhs[1];
//         };
//         priority_queue<vi, vvi, decltype(comp)> pq(points.begin(), points.end(), comp); // Heapify, O(n)

//         vvi ans;
//         while(K--) {
//             ans.push_back(pq.top()); 
//             pq.pop();
//         }
//         return ans;
//     }
// };

// class Solution {
// // QuickSelect by STL, Time: O(n), Space: O(1)
// public:
//     vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
//         auto comp = [](vi &lhs, vi &rhs) {
//             return lhs[0]*lhs[0]+lhs[1]*lhs[1] < rhs[0]*rhs[0]+rhs[1]*rhs[1];
//         };
//         nth_element(points.begin(), points.begin()+K-1, points.end(), comp);
//         return vvi(points.begin(), points.begin()+K);
//     }
// };

class Solution {
// QuickSelect, Time: O(n), Space: O(1)
public:
    int partition(vvi &points, int left, int right) {
        auto dist = [](vi &point) {
            return point[0]*point[0] + point[1]*point[1];
        };
        int i = left;
        int j = right;
        while(i != j) {          
            while(dist(points[j]) > dist(points[left]) && i < j) j--;
            while(dist(points[i]) <= dist(points[left]) && i < j) i++;
            if(i < j) swap(points[i], points[j]);
        }
        swap(points[left], points[i]);
        return i;
    }
    void quickSelect(vvi &points, int left, int right, int k) {
        if(left >= right) return;

        int pivot = rand() % (right - left + 1) + left;
        swap(points[left], points[pivot]);

        int pivotIdx = partition(points, left, right);
        if(pivotIdx == k) return;
        else if(pivotIdx > k) quickSelect(points, left, pivotIdx-1, k);
        else quickSelect(points, pivotIdx+1, right, k);
    }
    vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
        quickSelect(points, 0, points.size()-1, K-1);
        return vvi(points.begin(), points.begin()+K);
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vvi points; int K; vvi ans;

    points = {{1,3},{-2,2}};
    K = 1;
    ans = solution.kClosest(points, K);
    assert(ans==vvi({{-2,2}}));

    points = {{3,3},{5,-1},{-2,4}};
    K = 2;
    ans = solution.kClosest(points, K);
    // assert(ans==vvi({{-2,4},{3,3}}));
    assert(ans==vvi({{3,3},{-2,4}}));
}